### PR TITLE
[6.x] Prevent to serialize uninitialized properties

### LINKED ISSUE #33637: PHP7.4 only | Unhandled "Uninitialized" in serialization
- Laravel Version: 7.21.0
- PHP Version: 7.4.4

### Description:
The with PHP7.4 introduced Typed Properties V2 have also added a new state.

> If a typed property does not have a default value, no implicit null default value is implied (even if the property is nullable). Instead, the property is considered to be uninitialized. 

https://wiki.php.net/rfc/typed_properties_v2#uninitialized_and_unset_properties

If a typed property, which is initalized in the `handle()`-method, is used in a Job using `SerializesModels` trait, the job will break on Serialization with the Exception

> Typed property [class]::[property] must not be accessed before initialization

### Steps To Reproduce:

- Create a new Job with a typed property, which you do not initalize (see example below)
- Queue the Job
- You will get `Typed property [class]::[property] must not be accessed before initialization`-Exception

```php
class ProcessPodcast implements ShouldQueue
{
    use Dispatchable, InteractsWithQueue, Queueable, SerializesModels;

    // this property can be serialized, as it is initalized in the constructor (before serialize-deserialize)
    protected Podcast $podcast;

    // but when it tries to serialize this property to push the job to the queue, 
    // you will get the exception, as the handle-method has not run yet
    protected AudioProcessor $processor;

    public function __construct(Podcast $podcast)
    {
        $this->podcast = $podcast;
    }

    public function handle(AudioProcessor $processor)
    {
        $this->processor = $processor;

       // here's some very complex stuff which uses the processor in various methods, that's why it's made accessible as class-property.
    }
}
```

### Suggested Solution:

In PHP7.4, the corresponding method [ReflectionProperty::isInitialized](https://www.php.net/manual/en/reflectionproperty.isinitialized.php) was introduced. So for >= 7.4 we can just check this before accessing the the the property, for example in [SerializesModels:129](https://github.com/laravel/framework/blob/7.x/src/Illuminate/Queue/SerializesModels.php#L129). But I do not yet know how to resolve it in a backwards compatible way.

### Workaround:

Anyone finding this and in need of a workaround:
Create a nullable typed property instead and set a default value, like:
```php
protected ?AudioProcessor $processor = null;
```

### PR DESCRIPTION
**Pull request by dkulyk**
Fixes #33637

No need to serialize uninitialized properties.

### CODE REVIEW

**Comment by GrahamCampbell** [comment]
Could you add a test which covers this please?

**Review by GrahamCampbell — APPROVED** [review]

File: src/Illuminate/Queue/SerializesModels.php [hunk 1]
```diff
@@ -65,6 +65,12 @@ public function __serialize()
                 continue;
             }
 
+            $property->setAccessible(true);
+
+            if (! $property->isInitialized($this)) {
```
Comment (GrahamCampbell) [inline src/Illuminate/Queue/SerializesModels.php:70]: How come this code doesn't crash on PHP 7.2 and 7.3. Are you sure the tests cover it?

Comment (GrahamCampbell) [reply src/Illuminate/Queue/SerializesModels.php:70]: Don't you need to check the PHP version is at least 7.4 before calling that method?

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]

File: src/Illuminate/Queue/SerializesModels.php (hunk 1 above)
Comment (dkulyk) [reply src/Illuminate/Queue/SerializesModels.php:70]: function __serialize() works only on PHP 7.4+

Comment (GrahamCampbell) [reply src/Illuminate/Queue/SerializesModels.php:70]: Ah, yes. Thanks for clarifying.

**Review by GrahamCampbell — APPROVED** [review]

**Comment by taylorotwell** [comment]
This method only exists on PHP >= 7.4?

**Comment by dkulyk** [comment]
This method used by PHP >= 7.4 only

**Comment by driesvints** [comment]
I'd argue that the previous behavior was correct. The property is indicated as non-nullable and therefor serialization fails. If if needs to be nullable then you need to indicate it as such.

**Pull request merged**